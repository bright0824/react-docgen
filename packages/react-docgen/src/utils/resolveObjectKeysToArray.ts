import resolveToValue from './resolveToValue.js';
import type { NodePath } from '@babel/traverse';
import type {
  CallExpression,
  Identifier,
  NumericLiteral,
  StringLiteral,
} from '@babel/types';

function isObjectKeysCall(path: NodePath): path is NodePath<CallExpression> {
  if (!path.isCallExpression() || path.get('arguments').length !== 1) {
    return false;
  }

  const callee = path.get('callee');

  if (!callee.isMemberExpression()) {
    return false;
  }
  const object = callee.get('object');
  const property = callee.get('property');

  return (
    object.isIdentifier({ name: 'Object' }) &&
    property.isIdentifier({ name: 'keys' })
  );
}

function isWhitelistedObjectProperty(path: NodePath): boolean {
  if (path.isSpreadElement()) return true;

  if (
    path.isObjectProperty() ||
    (path.isObjectMethod() &&
      (path.node.kind === 'get' || path.node.kind === 'set'))
  ) {
    const key = path.get('key') as NodePath;

    return (
      (key.isIdentifier() && !path.node.computed) ||
      key.isStringLiteral() ||
      key.isNumericLiteral()
    );
  }

  return false;
}

function isWhiteListedObjectTypeProperty(path: NodePath): boolean {
  return (
    path.isObjectTypeProperty() ||
    path.isObjectTypeSpreadProperty() ||
    path.isTSPropertySignature()
  );
}

// Resolves an ObjectExpression or an ObjectTypeAnnotation
export function resolveObjectToNameArray(
  objectPath: NodePath,
  raw = false,
): string[] | null {
  if (
    (objectPath.isObjectExpression() &&
      objectPath.get('properties').every(isWhitelistedObjectProperty)) ||
    (objectPath.isObjectTypeAnnotation() &&
      objectPath.get('properties').every(isWhiteListedObjectTypeProperty)) ||
    (objectPath.isTSTypeLiteral() &&
      objectPath.get('members').every(isWhiteListedObjectTypeProperty))
  ) {
    let values: string[] = [];
    let error = false;
    const properties = objectPath.isTSTypeLiteral()
      ? objectPath.get('members')
      : (objectPath.get('properties') as NodePath[]);

    properties.forEach((propPath) => {
      if (error) return;

      if (
        propPath.isObjectProperty() ||
        propPath.isObjectMethod() ||
        propPath.isObjectTypeProperty() ||
        propPath.isTSPropertySignature()
      ) {
        const key = propPath.get('key') as
          | NodePath<Identifier>
          | NodePath<NumericLiteral>
          | NodePath<StringLiteral>;
        // Key is either Identifier or Literal
        const name: string = key.isIdentifier()
          ? key.node.name
          : raw
          ? (key.node.extra?.raw as string)
          : `${key.node.value}`;

        values.push(name);
      } else if (
        propPath.isSpreadElement() ||
        propPath.isObjectTypeSpreadProperty()
      ) {
        let spreadObject = resolveToValue(propPath.get('argument') as NodePath);

        if (spreadObject.isGenericTypeAnnotation()) {
          const typeAliasRight = resolveToValue(spreadObject.get('id')).get(
            'right',
          ) as NodePath;

          if (typeAliasRight.isObjectTypeAnnotation()) {
            spreadObject = resolveToValue(typeAliasRight);
          }
        }

        const spreadValues = resolveObjectToNameArray(spreadObject);

        if (!spreadValues) {
          error = true;

          return;
        }
        values = [...values, ...spreadValues];
      }
    });

    if (!error) {
      return values;
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the keys resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier keys
 */
export default function resolveObjectKeysToArray(
  path: NodePath,
): string[] | null {
  if (isObjectKeysCall(path)) {
    const objectExpression = resolveToValue(path.get('arguments')[0]);
    const values = resolveObjectToNameArray(objectExpression);

    if (values) {
      const nodes = values
        //filter duplicates
        .filter((value, index, array) => array.indexOf(value) === index)
        .map((value) => `"${value}"`);

      return nodes;
    }
  }

  return null;
}
