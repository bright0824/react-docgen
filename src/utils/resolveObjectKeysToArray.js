/*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import recast from 'recast';
import resolveToValue from './resolveToValue';

var {
  types: {
    ASTNode,
    NodePath,
    builders,
    namedTypes: types,
  },
} = recast;

function isObjectKeysCall(node: ASTNode): bool {
  return types.CallExpression.check(node) &&
    node.arguments.length === 1 &&
    types.MemberExpression.check(node.callee) &&
    types.Identifier.check(node.callee.object) &&
    node.callee.object.name === 'Object' &&
    types.Identifier.check(node.callee.property) &&
    node.callee.property.name === 'keys';
}

export function resolveObjectExpressionToNameArray(objectExpression: NodePath, raw: boolean = false): ?Array<string> {
  if (
    types.ObjectExpression.check(objectExpression.value) &&
    objectExpression.value.properties.every(
      prop =>
        types.Property.check(prop) &&
        (
          (types.Identifier.check(prop.key) && !prop.computed) ||
          types.Literal.check(prop.key)
        ) ||
        types.SpreadProperty.check(prop)
    )
  ) {
    let values = [];
    let error = false;
    objectExpression.get('properties').each(propPath => {
      if (error) return;
      const prop = propPath.value;

      if (types.Property.check(prop)) {
        // Key is either Identifier or Literal
        const name = prop.key.name || (raw ? prop.key.raw : prop.key.value);

        values.push(name);
      } else if (types.SpreadProperty.check(prop)) {
        const spreadObject = resolveToValue(propPath.get('argument'));
        const spreadValues = resolveObjectExpressionToNameArray(spreadObject);
        if (!spreadValues) {
          error = true;
          return;
        }
        values = [...values, ...spreadValues];
      }

    });

    if (!error) {
      return values;
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the keys resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier keys
 */
export default function resolveObjectKeysToArray(path: NodePath): ?NodePath {
  var node = path.node;

  if (isObjectKeysCall(node)) {
    const objectExpression = resolveToValue(path.get('arguments').get(0));
    const values = resolveObjectExpressionToNameArray(objectExpression);

    if (values) {
      const nodes = values
        .filter((value, index, array) => array.indexOf(value) === index)
        .map(value => builders.literal(value))

      return new NodePath(builders.arrayExpression(nodes));
    }
  }

  return null;
}
