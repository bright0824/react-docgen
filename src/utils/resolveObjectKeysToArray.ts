import {
  ASTNode,
  builders,
  NodePath as NodePathConstructor,
  namedTypes as t,
} from 'ast-types';
import type { NodePath } from 'ast-types/lib/node-path';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';

function isObjectKeysCall(node: ASTNode): boolean {
  return (
    t.CallExpression.check(node) &&
    node.arguments.length === 1 &&
    t.MemberExpression.check(node.callee) &&
    t.Identifier.check(node.callee.object) &&
    node.callee.object.name === 'Object' &&
    t.Identifier.check(node.callee.property) &&
    node.callee.property.name === 'keys'
  );
}

function isWhitelistedObjectProperty(prop) {
  return (
    (t.Property.check(prop) &&
      ((t.Identifier.check(prop.key) && !prop.computed) ||
        t.Literal.check(prop.key))) ||
    t.SpreadElement.check(prop)
  );
}

function isWhiteListedObjectTypeProperty(prop) {
  return (
    t.ObjectTypeProperty.check(prop) ||
    t.ObjectTypeSpreadProperty.check(prop) ||
    t.TSPropertySignature.check(prop)
  );
}

// Resolves an ObjectExpression or an ObjectTypeAnnotation
export function resolveObjectToNameArray(
  object: NodePath,
  importer: Importer,
  raw = false,
): string[] | null {
  if (
    (t.ObjectExpression.check(object.value) &&
      object.value.properties.every(isWhitelistedObjectProperty)) ||
    (t.ObjectTypeAnnotation.check(object.value) &&
      object.value.properties.every(isWhiteListedObjectTypeProperty)) ||
    (t.TSTypeLiteral.check(object.value) &&
      object.value.members.every(isWhiteListedObjectTypeProperty))
  ) {
    let values: string[] = [];
    let error = false;
    const properties = t.TSTypeLiteral.check(object.value)
      ? object.get('members')
      : object.get('properties');
    properties.each((propPath: NodePath) => {
      if (error) return;
      const prop = propPath.value;

      if (
        t.Property.check(prop) ||
        t.ObjectTypeProperty.check(prop) ||
        t.TSPropertySignature.check(prop)
      ) {
        // Key is either Identifier or Literal
        // @ts-ignore
        const name = prop.key.name || (raw ? prop.key.raw : prop.key.value);

        values.push(name);
      } else if (
        t.SpreadElement.check(prop) ||
        t.ObjectTypeSpreadProperty.check(prop)
      ) {
        let spreadObject = resolveToValue(propPath.get('argument'), importer);
        if (t.GenericTypeAnnotation.check(spreadObject.value)) {
          const typeAlias = resolveToValue(spreadObject.get('id'), importer);
          if (t.ObjectTypeAnnotation.check(typeAlias.get('right').value)) {
            spreadObject = resolveToValue(typeAlias.get('right'), importer);
          }
        }

        const spreadValues = resolveObjectToNameArray(spreadObject, importer);
        if (!spreadValues) {
          error = true;
          return;
        }
        values = [...values, ...spreadValues];
      }
    });

    if (!error) {
      return values;
    }
  }

  return null;
}

/**
 * Returns an ArrayExpression which contains all the keys resolved from an object
 *
 * Ignores setters in objects
 *
 * Returns null in case of
 *  unresolvable spreads
 *  computed identifier keys
 */
export default function resolveObjectKeysToArray(
  path: NodePath,
  importer: Importer,
): NodePath | null {
  const node = path.node;

  if (isObjectKeysCall(node)) {
    const objectExpression = resolveToValue(
      path.get('arguments').get(0),
      importer,
    );
    const values = resolveObjectToNameArray(objectExpression, importer);

    if (values) {
      const nodes = values
        .filter((value, index, array) => array.indexOf(value) === index)
        .map(value => builders.literal(value));

      return new NodePathConstructor(builders.arrayExpression(nodes));
    }
  }

  return null;
}
