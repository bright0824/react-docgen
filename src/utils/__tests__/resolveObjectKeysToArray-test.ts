import { builders } from 'ast-types';
import {
  statement,
  noopImporter,
  makeMockImporter,
  expressionLast,
} from '../../../tests/utils';
import resolveObjectKeysToArray from '../resolveObjectKeysToArray';

describe('resolveObjectKeysToArray', () => {
  const mockImporter = makeMockImporter({
    foo: statement(`
      export default {
        bar: "bar",
        "foo": "foo",
        1: 0,
        2: 5,
        [3]: 3,
        ['baz']: "baz",
      };
    `).get('declaration'),

    bar: statement(`
      export default {
        bar: 'bar',
      };
    `).get('declaration'),
  });

  it('resolves Object.keys with identifiers', () => {
    const path = expressionLast(
      ['var foo = { bar: 1, foo: 2 };', 'Object.keys(foo);'].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('bar'),
        builders.literal('foo'),
      ]),
    );
  });

  it('resolves Object.keys with literals', () => {
    const path = expressionLast(
      ['var foo = { "bar": 1, 5: 2 };', 'Object.keys(foo);'].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('bar'),
        builders.literal('5'),
      ]),
    );
  });

  it('resolves Object.keys with literals as computed key', () => {
    const path = expressionLast(
      ['var foo = { ["bar"]: 1, [5]: 2};', 'Object.keys(foo);'].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('bar'),
        builders.literal('5'),
      ]),
    );
  });

  it('resolves Object.keys when using resolvable spread', () => {
    const path = expressionLast(
      [
        'var bar = { doo: 4 }',
        'var foo = { boo: 1, foo: 2, ...bar };',
        'Object.keys(foo);',
      ].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('boo'),
        builders.literal('foo'),
        builders.literal('doo'),
      ]),
    );
  });

  it('resolves Object.keys when using getters', () => {
    const path = expressionLast(
      ['var foo = { boo: 1, foo: 2, get bar() {} };', 'Object.keys(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('boo'),
        builders.literal('foo'),
        builders.literal('bar'),
      ]),
    );
  });

  it('resolves Object.keys when using setters', () => {
    const path = expressionLast(
      [
        'var foo = { boo: 1, foo: 2, set bar(e) {} };',
        'Object.keys(foo);',
      ].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('boo'),
        builders.literal('foo'),
        builders.literal('bar'),
      ]),
    );
  });

  it('resolves Object.keys but ignores duplicates', () => {
    const path = expressionLast(
      [
        'var bar = { doo: 4, doo: 5 }',
        'var foo = { boo: 1, foo: 2, doo: 1, ...bar };',
        'Object.keys(foo);',
      ].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('boo'),
        builders.literal('foo'),
        builders.literal('doo'),
      ]),
    );
  });

  it('resolves Object.keys but ignores duplicates with getter and setter', () => {
    const path = expressionLast(
      ['var foo = { get x() {}, set x(a) {} };', 'Object.keys(foo);'].join(
        '\n',
      ),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toEqualASTNode(
      builders.arrayExpression([builders.literal('x')]),
    );
  });

  it('does not resolve Object.keys when using unresolvable spread', () => {
    const path = expressionLast(
      ['var foo = { bar: 1, foo: 2, ...bar };', 'Object.keys(foo);'].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toBeNull();
  });

  it('does not resolve Object.keys when using computed keys', () => {
    const path = expressionLast(
      ['var foo = { [bar]: 1, foo: 2 };', 'Object.keys(foo);'].join('\n'),
    );

    expect(resolveObjectKeysToArray(path, noopImporter)).toBeNull();
  });

  it('can resolve imported objects passed to Object.keys', () => {
    const path = expressionLast(`
      import foo from 'foo';
      Object.keys(foo);
    `);

    expect(resolveObjectKeysToArray(path, mockImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('bar'),
        builders.literal('foo'),
        builders.literal(1),
        builders.literal(2),
        builders.literal(3),
        builders.literal('baz'),
      ]),
    );
  });

  it('can resolve spreads from imported objects', () => {
    const path = expressionLast(`
      import bar from 'bar';
      var abc = { foo: 'foo', baz: 'baz', ...bar };
      Object.keys(abc);
    `);

    expect(resolveObjectKeysToArray(path, mockImporter)).toEqualASTNode(
      builders.arrayExpression([
        builders.literal('foo'),
        builders.literal('baz'),
        builders.literal('bar'),
      ]),
    );
  });
});
