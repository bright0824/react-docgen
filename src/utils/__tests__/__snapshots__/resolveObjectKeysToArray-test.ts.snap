// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveObjectKeysToArray can resolve imported objects passed to Object.keys 1`] = `
Array [
  "\\"bar\\"",
  "\\"foo\\"",
  "\\"1\\"",
  "\\"2\\"",
  "\\"3\\"",
  "\\"baz\\"",
]
`;

exports[`resolveObjectKeysToArray can resolve spreads from imported objects 1`] = `
Array [
  "\\"foo\\"",
  "\\"baz\\"",
  "\\"bar\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys but ignores duplicates 1`] = `
Array [
  "\\"boo\\"",
  "\\"foo\\"",
  "\\"doo\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys but ignores duplicates with getter and setter 1`] = `
Array [
  "\\"x\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys when using getters 1`] = `
Array [
  "\\"boo\\"",
  "\\"foo\\"",
  "\\"bar\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys when using resolvable spread 1`] = `
Array [
  "\\"boo\\"",
  "\\"foo\\"",
  "\\"doo\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys when using setters 1`] = `
Array [
  "\\"boo\\"",
  "\\"foo\\"",
  "\\"bar\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys with identifiers 1`] = `
Array [
  "\\"bar\\"",
  "\\"foo\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys with literals 1`] = `
Array [
  "\\"bar\\"",
  "\\"5\\"",
]
`;

exports[`resolveObjectKeysToArray resolves Object.keys with literals as computed key 1`] = `
Array [
  "\\"bar\\"",
  "\\"5\\"",
]
`;
